File: app/src/main/java/io/mgba/Adapters/Interfaces/MultipleViewHolderAdapter.java
package io.mgba.Adapters.Interfaces;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import java.util.HashMap;
import java.util.List;

import io.mgba.mgba;
import io.reactivex.functions.BiFunction;

public abstract class MultipleViewHolderAdapter<T extends IViewHolderCreator> extends BaseAdapter<T> {

    private final HashMap<Integer, Integer> layoutRouter;
    private final BiFunction<Integer, View, RecyclerView.ViewHolder> holderCreator;

    @Override
    public int getItemViewType(int position) {
        final IViewHolderCreator item = items.get(position);
        return item.getID();
    }


    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int layout = layoutRouter.get(viewType);
        View view = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);

        RecyclerView.ViewHolder holder = null;

        try {
            holder = holderCreator.apply(layout, view);
        } catch (Exception e) {
            mgba.report(e);
        }

        return holder;
    }

    public MultipleViewHolderAdapter(@NonNull Context context, @NonNull RecyclerView recyclerView,
                                     @NonNull HashMap<Integer, Integer> layoutRouter,
                                     @NonNull BiFunction<Integer, View, RecyclerView.ViewHolder> holderCreator) {
        super(context, recyclerView);
        this.layoutRouter = layoutRouter;
        this.holderCreator = holderCreator;
    }

    public MultipleViewHolderAdapter(@NonNull List<T> items, @NonNull RecyclerView recyclerView, @NonNull Context context,
                                     @NonNull HashMap<Integer, Integer> layoutRouter,
                                     @NonNull BiFunction<Integer, View, RecyclerView.ViewHolder> holderCreator) {
        super(items, recyclerView, context);
        this.layoutRouter = layoutRouter;
        this.holderCreator = holderCreator;
    }
}

File: app/src/main/java/io/mgba/adapters/interfaces/MultipleViewHolderAdapter.kt
package io.mgba.adapters.interfaces

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.util.HashMap
import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView

abstract class MultipleViewHolderAdapter<T : IViewHolderCreator> : BaseAdapter<T> {

    private val layoutRouter: HashMap<Int, Int>
    private val holderCreator: (Int, View) -> RecyclerView.ViewHolder

    override fun getItemViewType(position: Int): Int {
        val item = items[position]
        return item.id
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layout = layoutRouter[viewType]
        val view = LayoutInflater.from(parent.context).inflate(layout!!, parent, false)

        return holderCreator.invoke(layout, view)
    }

    constructor(@NonNull context: Context, @NonNull recyclerView: RecyclerView,
                @NonNull layoutRouter: HashMap<Int, Int>,
                @NonNull holderCreator: (Int, View) -> RecyclerView.ViewHolder) : super(emptyList(), recyclerView, context) {
        this.layoutRouter = layoutRouter
        this.holderCreator = holderCreator
    }

    constructor(@NonNull items: List<T>, @NonNull recyclerView: RecyclerView, @NonNull context: Context,
                @NonNull layoutRouter: HashMap<Int, Int>,
                @NonNull holderCreator: (Int, View) -> RecyclerView.ViewHolder) : super(items, recyclerView, context) {
        this.layoutRouter = layoutRouter
        this.holderCreator = holderCreator
    }
}
